src/main/java/com/ibm/disni/rdma/verbs/impl/NatPollCqCall.java

```diff
@@ -61,27 +61,25 @@
 	private int result;
 	private boolean valid;
 	
-	public NatPollCqCall(RdmaVerbsNat verbs, NativeDispatcher nativeDispatcher, MemoryAllocation memAlloc) {
+	public NatPollCqCall(RdmaVerbsNat verbs, NativeDispatcher nativeDispatcher,
+	                     MemoryAllocation memAlloc, IbvCQ cq, IbvWC[] wcList, int ne) {
 		this.verbs = verbs;
 		this.nativeDispatcher = nativeDispatcher;
 		this.memAlloc = memAlloc;
 		this.valid = false;
-	}
-
-	public void set(IbvCQ cq, IbvWC[] wcList, int ne) {
 		this.cq = (NatIbvCQ) cq;
 		this.wcList = wcList;
 		this.ne = ne;
-		
-		this.csize = wcList.length*IbvWC.CSIZE;
-		if (cmd != null){
-			cmd.getBuffer().clear();
-		} else {
-			this.cmd = memAlloc.allocate(csize);
-		}
 
+		this.csize = wcList.length*IbvWC.CSIZE;
+		this.cmd = memAlloc.allocate(csize);
 		this.valid = true;
 	}
+
+	public void setWcList(IbvWC[] wcList) {
```

Comment:
This call doesn't need to be public, does it?

Reply:
But i want to cache this objects on app level. Then i have something like this:
```java
 private final ConcurrentSkipListMap<Integer, NatPostRecvCall> postRecvCallCache =
    new ConcurrentSkipListMap<>();
NatPostRecvCall svcPostRecv = postRecvCallCache.remove(newRecvWrList.size());
      if (svcPostRecv == null) {
        svcPostRecv = (NatPostRecvCall)qp.postRecv(newRecvWrList, null);
      } else {
        svcPostRecv.setWrList(newRecvWrList);
      }
      svcPostRecv.execute();
```

Reply:
To cache objects on app level, to avoid buffers allocs. Then just update with a new wrList

Reply:
That's not the intention of SVC objects, the intention of SVC is objects is that applications can create them, cache them as is, and then use the modifying operations on them to change the offset etc. If the application needs a new wrList then the application should just create a new SVC object. SVC objects are defined to be immutable with respect to the underlying memory size. Having a call that appears to allow users to set a completely new wrlist, only to then fail the assert if the new wrlist is greater I consider to be bad practice.

Reply:
Way for applications to use SVC objects:

1. Create SVC object
2. Use it to issue verbs call
3. Cache it
4. Take it out of the cache and prepare it for a new operation: 
Modify it using the modifying methods (e.g., setRemote_addr)
5. Issue verbs call
6. Loop 4-5

Reply:
But in this case need to expose `sgeNatList` `wrNatList`  and `cmd` to update buffer on application level. Since if just to expose `getWrList()` method, update wrList on application level, it won't propagate to underlining byte buffer.

Reply:
setRemote_addr does propagate to underlying byte buffer, and so are also the other calls (setRkey, setWr_id, etc.)

Reply:
For an example of how to use SVCs have a look at this:
https://github.com/zrlio/disni/blob/master/src/test/java/com/ibm/disni/examples/ReadClient.java#L78

Reply:
I mean these calls to make at application level:
```java
for (NatIbvSendWR natWR : wrNatList){
			natWR.writeBack(cmd.getBuffer());
}
for (NatIbvSge sge : sgeNatList){
	sge.writeBack(cmd.getBuffer());
}
```

Reply:
In that example `endPoint.postSend` will come up to `qp.postSend` that was call `set` method:
https://github.com/zrlio/disni/blob/76e4ae439b12e995922744b72b150f61fd551e18/src/main/java/com/ibm/disni/rdma/RdmaEndpoint.java#L295

Reply:
Can't follow. What are you saying?

Reply:
Let's look at at [NatPostSendCall](https://github.com/zrlio/disni/blob/master/src/main/java/com/ibm/disni/rdma/verbs/impl/NatPostSendCall.java):
1. In `execute` we pass a buffer that has NatWrItems and NatSGItems, that was set here:
https://github.com/zrlio/disni/blob/master/src/main/java/com/ibm/disni/rdma/verbs/impl/NatPostSendCall.java#L99-L104
2. I've cached this object, based on wrList size, now i want to update it's wrList with *the SAME* wrlist size.
3. Need somehow to update object's buffer to call execute with correct data.

Reply:
Example: Let's say you app periodically issues read against RDMA server for different memory areas, Then your app may want to create N different PostSendCalls and cache them, and every time the app does a call it takes the cached SVC for that particular server out of the cache, adjusts the offset using the modifiers, and isues the read.

Reply:
The code above you are showing anyway is a problem anyway as it operates on the NatPostRecvCall type, which is an implementation we don't want to export (with Java 9 we want to hide that type completely so that's in not accessible). The exported type for this PostRecvCall which does not export the set method.

Reply:
Generally, if you need a type cast then you're doing something wrong.

Reply:
Ok updated. The issue is that there's a lot of buffer allocations in hot path. The best way is to simply cache more buffers (currently we cache only 1 buffer per size):

![image](https://user-images.githubusercontent.com/1121987/43576928-27b8babe-9653-11e8-94ca-46482d977bbe.png)